gtk/src/toga_gtk/hardware/location.py

```diff
@@ -0,0 +1,146 @@
+from enum import IntEnum, auto
+
+import gi
+
+from toga import LatLng
+from toga_gtk.libs import Gio, GLib, GObject
+
+gi.require_version("Geoclue", "2.0")
+from gi.repository import Geoclue  # noqa: E402
+
+
+def toga_location(location):
+    """Convert a ``Geoclue.Location`` into ``OnLocationChangeHandler`` kwargs."""
+    latlng = LatLng(*location.get_properties("latitude", "longitude"))
+    altitude = location.get_property("altitude")
+
+    return {
+        "location": latlng,
+        "altitude": altitude,
+    }
+
+
+class States(IntEnum):
+    STARTING = auto()
+    """Waiting for the Geoclue client to initialise."""
+
+    READY = auto()
+    """Geoclue is ready to connect."""
+
+    MONITORING = auto()
+    """Actively monitoring the location."""
+
+    FAILED = auto()
+    """``Geoclue`` was unable to retrieve the location due to a generic error."""
+
+    DENIED = auto()
+    """Permission to read the location was denied."""
+
+
+class Location(GObject.Object):
+    # Implement as GObject.Object to allow notifying on state changes within the event loop
+    state = GObject.Property(type=int, default=States.STARTING)
+
+    def __init__(self, interface):
+        super().__init__()
+
+        self.interface = interface
+
+        self.geoclue_simple = None
+        self.notify_location_handle = None
+        self.notify_active_listener = None
+        self.in_flatpak = None
```

Comment:
This is an ugly `Optional[bool]`. I might refactor this to an enum `Sandboxed` so it can either be "undetermined", "sandboxed", or "not sandboxed". It also doesn't need to really be explicitly _Flatpak_ here, though that's the only relevant case I'm aware of for the time being.

Reply:
You aren't missing a case, though, to clarify. The cases as far as permissions go are:

- Sandboxed
  - Permission denied? Toga gets a `FAILED` error
  - Permission granted? It just works
- Unsandboxed
  - No such thing as permissions, location request is always allowed. [You can read more about that here on the GNOME forums](https://discourse.gnome.org/t/geoclue2-local-user-security-model/9996/2).

> Aside on error handling...
>
> If you try getting a location and e.g., the WiFi location service (MLS in this case) doesn't work, then `new_finish` will never run. It just hangs indefinitely, and there's no real way to tell what's going on. If you search for Geoclue issues around the web, you'll see lots of people reporting that kind of behaviour with redshift in particular, but also GNOME Maps and Weather.
>
> In which case, there's another case where regardless of sandboxing status, Geoclue never responds. This is basically the reason why tracking state as initial vs starting is useful... in the future Toga GTK could emit a useful error if `Location` does not finish starting within a timeout period.

`Geoclue.Simple` will automatically use the [XDG Location portal](https://flatpak.github.io/xdg-desktop-portal/docs/doc-org.freedesktop.portal.Location.html) when inside a Flatpak. I don't know how snap sandboxing works, I presume it would be the same. Outside a sandboxed environment, the DBus API is used directly. This behaviour of `Geoclue.Simple` is also relied upon by GNOME's apps like Maps and Weather.

From what I understand, `Gtk.FileChooserDialog` has similar behaviour and will use the XDG portal when sandboxed, or the DBus API directly otherwise.

A Flatpak app does not need to pre-declare special permissions for location usage. Upon starting `Geoclue.Simple` and accessing the XDG portal, the user is prompted for permission. `.desktop` files appear to support `X-Geoclue-Reason`, or... at least GNOME Maps has this key, but I can't find documentation about it so far, nor have I seen this text show up when I try it for myself. Maybe it is only supported by GNOME.

The last piece of this puzzle for me is that I'm still not 100% clear on how to interact with these permission APIs directly to tell if permission has been _rejected_, or whether that's necessary to keep track of. As far as I can tell from the PyGObject docs, there is no `Gio.Permission` class for location permission. But, XDG Portal has a `PermissionStore` that you can request permissions from: https://flatpak.github.io/xdg-desktop-portal/docs/doc-org.freedesktop.impl.portal.PermissionStore.html#org-freedesktop-impl-portal-permissionstore-lookup

The testbed has a distinction between "allowing" location requests and "granting" location requests, and I'm having some trouble wrapping my head around this and how it needs to work for Toga GTK without making Toga GTK's location tests a weird edge case.

Reply:
> In which case, there's another case where regardless of sandboxing status, Geoclue never responds. This is basically the reason why tracking state as initial vs starting is useful... in the future Toga GTK could emit a useful error if `Location` does not finish starting within a timeout period.

iOS, macOS and Android requests can also fail (e.g., when you're sitting in your Faraday cage on the bottom floor of your volcano lair). So if there's a convenient way to catch timeouts, that would be good sounds like a good addition to the GTK backend - but that would be it's a failure of a specific request, not a failure of permissions. Permission is a guarantee that you can *make* the request, not a guarantee that it will succeed.
 
> A Flatpak app does not need to pre-declare special permissions for location usage. Upon starting `Geoclue.Simple` and accessing the XDG portal, the user is prompted for permission. `.desktop` files appear to support `X-Geoclue-Reason`, or... at least GNOME Maps has this key, but I can't find documentation about it so far, nor have I seen this text show up when I try it for myself. Maybe it is only supported by GNOME.

Ok - sounds like adding this to the .desktop file might be worth adding as a nice-to-have, but not worth stressing about too much.
 
> The last piece of this puzzle for me is that I'm still not 100% clear on how to interact with these permission APIs directly to tell if permission has been _rejected_, or whether that's necessary to keep track of. As far as I can tell from the PyGObject docs, there is no `Gio.Permission` class for location permission. But, XDG Portal has a `PermissionStore` that you can request permissions from: https://flatpak.github.io/xdg-desktop-portal/docs/doc-org.freedesktop.impl.portal.PermissionStore.html#org-freedesktop-impl-portal-permissionstore-lookup

That's somewhat of a platform question - iOS, macOS and Android can all tell you explicitly "you do not have permission", so there's no need to explicitly store a "rejected" state - you can determine the current state from the platform. 

If there's no direct "do I have permission" API for GTK (which sounds like might be the case), I'd be fine with a "speculative location request" as a proof of permission - i.e., `request_permission` actually tries to get the permission; if it succeeds, you have permission, and you discard (or maybe cache) the result; if the request fails with an error, you don't have permission.

This actually plays into something that occurred to me in the process of writing my PyCon AU talk - while the API is set up so that you have to request permission before using the API... I can't think of a good reason why that request can't be implicit on the API calls themselves - ie., you "just call" a location API, and if you haven't got permission, permission is requested. 

This has a couple of benefits:
* We'd be making the API easier to use (one less call required)
* It's more obvious handling for the "permission change at runtime" case - you don't need to put "request" calls before every "use" call.
* This is the baked in behaviour on iOS anyway - so the implementation would be simpler (on iOS, anyway), and not that much more complex everywhere else.
* We'd also be improving adherence with App Store guidelines -  both iOS and Android agree that permission requests should be occurring as close to the first request as possible, so that there's clear connection between the request and the action.

So - if removing the prominence of the `request_permission` API would make the Linux implementation easier, I'm not fundamentally opposed to that.
 
> The testbed has a distinction between "allowing" location requests and "granting" location requests, and I'm having some trouble wrapping my head around this and how it needs to work for Toga GTK without making Toga GTK's location tests a weird edge case.

The secret here is the difference between the current permission, and the future state permission. There are 3 types of method:
* **grant**: set up the test so that current permission status is ALLOWED
* **reject**: set up the test so that current permission status is DENIED
* **allow**: set up the test so current permission status is UNKNOWN, but if the user requests permission, it will be ALLOWED

These then exist for both foreground and background permissions.

Reply:
Awesome, thanks as always for the thorough response!

> If there's no direct "do I have permission" API for GTK (which sounds like might be the case), I'd be fine with a "speculative location request" as a proof of permission - i.e., request_permission actually tries to get the permission; if it succeeds, you have permission, and you discard (or maybe cache) the result; if the request fails with an error, you don't have permission.

After much poking around, the XDG Portal PermissionStore _is_ there, but it's on the `org.freedesktop.impl.portal` bus namespace, and sandboxed apps only have access to `org.freedesktop.portal` namespace (no `.impl`). In other words, the only time it's relevant, it isn't available.

A speculative location request is, in the end, necessary. Incidentally, that's what I implemented in this commit on `Location._request_permission`: https://github.com/beeware/toga/pull/2999/commits/0e0ec80c4e03146e3f4678c29852c2ba48017cf8#diff-7003a350ef59818bcbe863c85ad516382a8e5ae5fcb8eb0bb09fe639e5843e60R142-R143

> The secret here is the difference between the current permission, and the future state permission. [...]
> 
> These then exist for both foreground and background permissions.

This was a perfect explanation, thank you! The tests are all passing now except `test_deny_background_permission`. Because there is no distinction between foreground and background permissions, I'm not sure how the GTK `Location` could or should track a denial of _just_ the background permissions. There's no meaningful distinction for this as far as location goes... but there is the XDG Portal "background" permission for apps that want to run in the background. I could see interpreting background location requests as requiring the XDG Portal background permission... but I'd need to understand the use case of that on GTK.

I suspect the answer for that will come from the mobile Linux space. I'll take a look and see if Phosh has this distinction.

Reply:
On a separate note, regarding the permission request timing and this note:

> This actually plays into something that occurred to me in the process of writing my PyCon AU talk - while the API is set up so that you have to request permission before using the API... I can't think of a good reason why that request can't be implicit on the API calls themselves - ie., you "just call" a location API, and if you haven't got permission, permission is requested.

I think that would be a good idea. I'm not sure if it's necessary for GTK, but it would certainly simplify the implementation for the GTK interface, considering the permission request does indeed happen as part of requesting the location for the first time, rather than as a distinct and separate request.

No need for big refactors to accommodate GTK, but worth pointing out that it'll probably be a nice cleanup and simplification of the GTK `Location` implementation, similar to iOS, it sounds like.

Reply:
> After much poking around, the XDG Portal PermissionStore _is_ there, but it's on the `org.freedesktop.impl.portal` bus namespace, and sandboxed apps only have access to `org.freedesktop.portal` namespace (no `.impl`). In other words, the only time it's relevant, it isn't available.

🤦 

> The tests are all passing now except `test_deny_background_permission`.  Because there is no distinction between foreground and background permissions, I'm not sure how the GTK `Location` could or should track a denial of _just_ the background permissions. There's no meaningful distinction for this as far as location goes... but there is the XDG Portal "background" permission for apps that want to run in the background. I could see interpreting background location requests as requiring the XDG Portal background permission... but I'd need to understand the use case of that on GTK.
> 
> I suspect the answer for that will come from the mobile Linux space. Tomorrow I will take a look and see if Phosh has this distinction. Otherwise, I'll need some advice for what to do with this distinction not existing for GTK.

If you're able to make this work, that's great - but if you can't, I'd be entirely comfortable with essentially removing the distinction - treating "location permission" as a single thing, so if you have "foreground" permission, you also get background permission; or, at the very least, that you get background permission as soon as you ask for foreground permission. The latter may be necessary for consistency in the API - other platforms won't let you ask for background permission without having foreground permission (and you can't *only* have background permission), so we might need to enforce that restriction even if it doesn't *really* exist on GTK.

If you *do* come across a test that can't be reproduced because it doesn't make sense under GTK circumstances, it's OK to skip that test. We generally do that by making the probe method that is needed to set up or measure the feature raise a `skip` or `xfail` (depending on whether the feature is "not implemented yet" or "fundamentally doens't make sense on the platform and *won't* be implemented"). Alternatively, if there's no obvious probe method that can be used, we'll add a boolean property on the probe (e.g., `supports_background_location` in this case), and then perform a skip or xfail based on that.

Reply:
> If you do come across a test that can't be reproduced because it doesn't make sense under GTK circumstances, it's OK to skip that test. We generally do that by making the probe method that is needed to set up or measure the feature raise a skip or xfail (depending on whether the feature is "not implemented yet" or "fundamentally doens't make sense on the platform and won't be implemented"). Alternatively, if there's no obvious probe method that can be used, we'll add a boolean property on the probe (e.g., supports_background_location in this case), and then perform a skip or xfail based on that.

Aha! I figured out (and used) the `xfail` approach for the location rejection test, having read the Android tests. I did not see other examples of skipping the test using the boolean, so it might just be necessary to do that. As it stands, it would delete a few of the more ugly bits of code if background permission was not treated as distinct in GTK.

We'll see what I can find in Phosh tomorrow :slightly_smiling_face:

Reply:
I've resolved this question by pushing off considering background permission in sandboxed apps for a separate issue. This PR is already complex enough as it is, and as far as `Geoclue.Simple` and XDG Location Portal are concerned, there is no difference between "tracking" location and doing a one-off read of the location.